WIP - Add facebook folly TDigest

mehdix300 opened the pull request.

Commit:
Add facebook folly TDigest

Commit:
To remove Range from TDigest body

Commit:
Update accumulators and SConscript

stephan-hof commented on src/mongo/db/pipeline/TDigest.cpp:
```diff
@@ -106,7 +108,7 @@ TDigest::TDigest(
 // possible.  This implementation puts all additional memory in the
 // heap, so that if called from fiber context we do not smash the
 // stack.  Otherwise it is very similar to boost::spreadsort.
-TDigest TDigest::merge(Range<const double*> unsortedValues) const {
```
I would say we get rid of this method completely.
We won't do the sorting in this class. It will always happen outside.

stephan-hof commented on src/mongo/db/pipeline/TDigest.cpp:
```diff
@@ -118,12 +120,12 @@ TDigest TDigest::merge(Range<const double*> unsortedValues) const {
 }
 
-TDigest TDigest::merge(presorted_t, Range<const double*> sortedValues) const {
+TDigest TDigest::merge(presorted_t, const std::vector<double> & sortedValues) const {
```
If the other method is removed we can simplify the signature here a little to
`TDigest TDigest::merge(const std::vector<double> &sortedValues) const`

So we get rid of this `presorted_t` type.

stephan-hof commented on src/mongo/db/pipeline/TDigest.cpp:
```diff
@@ -232,8 +234,8 @@ TDigest TDigest::merge(Range<const TDigest*> digests) {
     if (curCount > 0) {
-      DCHECK(!std::isnan(it->min_));
-      DCHECK(!std::isnan(it->max_));
+      //DCHECK(!std::isnan(it->min_));
```
Could you explain why this `DCHECK` have been removed ?

stephan-hof commented on src/mongo/db/pipeline/TDigest.cpp:
```diff
@@ -14,13 +14,15 @@
-#include <folly/stats/TDigest.h>
-#include <folly/stats/detail/DoubleRadixSort.h>
+#include "TDigest.h"
+#include "DoubleRadixSort.h"
```
I think you can remove this include of 'DoubleRadixSort.h' right ?

stephan-hof commented on src/mongo/db/pipeline/TDigest.h:
```diff
@@ -19,10 +19,10 @@
 
-#include <folly/Range.h>
-#include <folly/Utility.h>
+struct presorted_t {};
```
Now that we changed the 'merge' signature a little to only accept presorted values those two definitions can go away.

stephan-hof commented on src/mongo/db/pipeline/TDigest.h:
```diff
@@ -93,14 +93,14 @@ class TDigest {
    * Returns a new TDigest constructed with values merged from the current
    * digest and the given sortedValues.
    */
-  TDigest merge(presorted_t, Range<const double*> sortedValues) const;
```
Same as above. I would suggest we only need this merge method.
`TDigest merge(const std::vector<double>& sortedValues) const;`

stephan-hof reviewed (changes requested):
As written in the inline comments, lets remove the method which accepts unsorted values.

Commit:
adjust TDigest sortedvalues and create new accumulator

Commit:
Connect the new accumulator to TDigest

spencerjackson committed:
SERVER-38556 Handle transaction events in RoleGraph update

chasinglogic committed:
SERVER-32295 Support Python 3

henrikedin committed:
SERVER-39025 Fix feature subset installation from MSI. Missing yaml file is no longer critical.

Commit:
SERVER-39311 Enable burn_in repetition count: at least 2 executions, with a maximum of 1000 excutions or 10 minutes of execution time

Commit:
SERVER-40209 Implement $$NOW and $$CLUSTER_TIME

antkorsh committed:
SERVER-24860 Optimize away entire pipeline if it can be answered using a query

TedTuckman committed:
SERVER-39257 Implement visitor pattern for UpdateNode

bcaimano committed:
SERVER-39814 Add OutOfLineExecutor to ConnectionPool factory interface

bcaimano committed:
SERVER-39814 Add OutOfLineExecutor to ConnectionPool factory interface

bcaimano committed:
SERVER-39965 OutOfLineExecutor Tasks are now unique_function(Status)

bcaimano committed:
SERVER-39965 OutOfLineExecutor Tasks are now unique_function(Status)

spencerjackson committed:
SERVER-39825 Add ESE sys-perf variants

Commit:
SERVER-39313 Metric tracking script for burn_in_tests

hanumantmk committed:
SERVER-35114 Adjustable periods for PeriodicJobs

Add support for changing the period of a periodic job

vessy-mongodb committed:
SERVER-39989 Use a config.transactions find command for the begin fetching optime in initial sync

puppyofkosh committed:
SERVER-13298 distinct now uses index for multikey dotted fields

shreyaskalyan committed:
SERVER-39895 Write shell JS API for interacting with key vault

jseyster committed:
SERVER-39676 Replace DBDirectClient in TransactionHistoryIterator

spencerjackson committed:
SERVER-40133 Add version readme for third party libraries

acmorrow committed:
SERVER-40179 Require tvOS 11 for embedded to allow use of sized allocation

Commit:
SERVER-39852 show "empty" databases as "(empty)"

* databases with reported non-zero sizes will be displayed
  as they always have, with their size displayed after the name
* databases that the server reports as empty, which generally
  indicates that it contains no collections, will be decorated
  with "(empty)"
* databases that the server reports as taking up 0 bytes will
  now listed without any decoration

jbreams committed:
SERVER-39820 Include client IP in log message for successful authentication

banarun committed:
SERVER-37848 Test $regex agg expressions with regex that meets and exceeds the capture limit

louiswilliams committed:
SERVER-39762 Fix fastcount after rollback recovery of prepared transactions

Commit:
SERVER-40312 Create a generic tree for pipeline metatdata

Commit:
SERVER-40312 Create a generic tree for pipeline metatdata

nzolnierzmdb committed:
SERVER-39248 Implement FCV check for new JSON Schema encryption keywords

banarun committed:
SERVER-40140 Fix Coverity analysis defect: Improper use of negative value

jannaerin committed:
SERVER-40200 remove query_exec lib dependency on transaction library

puppyofkosh committed:
SERVER-40246 Add cursor manager sys perf test

visualzhou committed:
SERVER-40081 Move session checkout to before command execution.

renctan committed:
SERVER-39175 Transactional bulkWrite error missing writeErrors (mongos)

Commit:
SERVER-39152 apply JSGCConfig fix from bz#1515475

Commit:
SERVER-39152 more correctly set JSGC_MAX_BYTES

Commit:
SERVER-39079 Move BackgroundOperation checks out of the catalog layer; add parallel IndexBuildsCoordinator checks for all BackgroundOperation checks

Commit:
SERVER-39831 Never update commit point beyond last applied if learned from sync source

renctan committed:
SERVER-37734 Add a default log component verbosity for mongos to resmoke, and set the 'transaction' component to 3

Commit:
SERVER-40292 Jstests that pass their own ReplSetTest to RollbackTest must disable chaining

spencerjackson committed:
SERVER-40242 Update banner inclusions for vendored tools

vrachev committed:
SERVER-39899 Enable the initial sync fuzzer in Evergreen

cswanson310 committed:
SERVER-40305 Add logging in max_time_ms.js

xy24 committed:
SERVER-39519 Create/drop view now only takes database IX lock

ajdavis committed:
SERVER-40307 Fix replSetInitiate retry logic

jseyster committed:
SERVER-39768 API to check if Stitch Support objects need a matcher

This includes stitch_support_v1_projection and stitch_support_v1_update.

BillyDonahue committed:
SERVER-40260 Backport gperftools upstream 62c4eca6e7e to gperftools-2.5.

Under x64, the PE loader looks for callbacks in constant sections

GWlodarek committed:
SERVER-35394 Ensure that only the internal role can run the prepareTransaction command

GWlodarek committed:
SERVER-35394 Test access control of prepareTransaction command

lukech committed:
Import wiredtiger: 67066a329d365cb1de6e198184016d4a9d280679 from branch mongodb-4.2

ref: bb36cf83cd..67066a329d
for: 4.1.10

WT-4324       Ensure checkpoints rewrite pages with data in the future
WT-4609       Extend timestamp usage documentation
WT-4621       Support rounding up prepare timestamp to oldest
WT-4629       Research timestamping and add to test coverage for timestamping
WT-4630       Add Clang Format script
WT-4645       Fill gaps in test coverage regarding transactions
WT-4646       Coverity doesn't like testutil_assert() on fopen calls
WT-4648       Coverity false positive on fopen call
WT-4651       Remove 'incase' from s_string.ok and its uses
WT-4654       Split task for Evergreen Windows build variant to reduce makespan
WT-4655       Avoid C and C++ compiler incompatibilities

BlakeIsBlake committed:
SERVER-40264 Change MigrationChunkCloner to be a shared pointer to synchronize object lifetime across threads

saltzm committed:
SERVER-39631 Move failpoint activation in lagged_config_secondary.js

jasonjhchan committed:
SERVER-39792 Update the txn table only on the first txn operation on primary

chasinglogic committed:
SERVER-40332 Add psutil to lint requirements

markbenvenuto committed:
SERVER-40132 Create push tasks for mongocryptd tgz

ajdavis committed:
SERVER-40069 Fix global lock tracking for txns

Includes SERVER-40084, stores global lock acquisition flags in an
atomic.

will62794 committed:
SERVER-39660 Return SnapshotUnavailable error when trying to read behind pending catalog changes in AutoGetCollectionForRead using a provided timestamp read source

BillyDonahue committed:
SERVER-40334 change --allocator=tcmalloc to mean gperftools-2.7

henrikedin committed:
SERVER-39728 Deinit SSL Manager on Windows in unittests.

This prevents the background threads being started in the Windows thread
pool after we return from main() while we're tearing down the CRT causing
access violations even though we returned 0 from main.

Commit:
SERVER-40271: Add TLS-enabled, 3-node replica set build variant to sys-perf

renctan committed:
SERVER-40301 Don't hold mutex while doing queries in nextCloneBatch and nextModsBatch

jsmulrow committed:
SERVER-39477 random_moveChunk workloads should tolerate failed migrations because chunk's range is in use on recipient

spencerjackson committed:
SERVER-40344 Upgrade ASAN builders to Ubuntu 18.04

lankas committed:
SERVER-39752 Unblacklist tests that were waiting for commitTransaction oplog entries to actually be applied

jimoleary committed:
SERVER-38914 raise setup failures for rejected tasks

TedTuckman committed:
SERVER-39257 Move FieldRefTempAppend to FieldRef class

henrikedin committed:
SERVER-32709 Improve performance with the mobile storage engine.

SQLite is configured with synchronous=NORMAL for all sessions.
Reduced temporary memory allocations when constructing SQL statements.
Implemented waitUntilDurable when this storage engine is used in mongod only.

BlakeIsBlake committed:
SERVER-39840 Add kExecutedRetryableWrite state to allow unstarted retryable writes to "upgrade" to transactions

mtrussotto committed:
SERVER-39443 Apply prepared transactions on secondary with the new > 16MB transaction oplog format.

jbreams committed:
SERVER-40296 Allow returning invalid users from authorization manager

mtrussotto committed:
SERVER-39444 Include prepared transaction tests in new passthrough test suites for large transactions

BlakeIsBlake committed:
SERVER-40333 Only clone collection options once for each shard in shardsvrShardCollection

henrikedin committed:
SERVER-39102 Fix embedded and stitch unittests so logging and time reporting works correctly.

Also fixed so embedded::initialize leaves a cleaner state in case of failure so initialize with correct parameters can be done afterwards.

Commit:
SERVER-40210 Integration tests for $$NOW/$$CLUSTER_TIME in find, aggregate, and views

Commit:
SERVER-40265 build mongoed with wiredtier

* Also, updated a couple of KVStorageEngine::TimestampMonitor tasks to
  not create a opCtx if it already exists

louiswilliams committed:
SERVER-40227 Pass ignore prepare behavior from a command attribute rather than using a whitelist

Commit:
SERVER-40212 Test $$NOW/$$CLUSTER_TIME in a sharded configuration

spencerjackson committed:
SERVER-40226 Upgrade static OpenSSL to 1.1.1b

benety committed:
SERVER-40347 optime.js determines storage engine support for persistence using server status

benety committed:
SERVER-40347 optime.js waits for durable wall clock times to synchronize across replica set nodes

kaloianm committed:
SERVER-40297 Move the TransactionCoordinator data structures to a separate file

TedTuckman committed:
SERVER-39254 Add getter to RenameNode

jbreams committed:
SERVER-40356 Make ConstDataRange constructable from any byte-like type

BillyDonahue committed:
SERVER-40357 expand all calls to MONGO_DISALLOW_COPYING

produced by:

hits="$(git grep -n MONGO_DISALLOW_COPYING | cut -d: -f1 )"
for f in "$hits"; do
    sed -i.orig '
    s/^\( *\)MONGO_DISALLOW_COPYING(\(.*\));/\1\2(const \2\&) = delete;\n\1\2\& operator=(const \2\&) = delete;/;
    ' $f
done

ajdavis committed:
SERVER-40355 Handle RS config with _id larger than set size

chasinglogic committed:
SERVER-40359 Bypass compile should run pip install

jasonjhchan committed:
SERVER-40247 Add new replica_sets_multi_oplog_txns test suite for large transactions

saltzm committed:
SERVER-37974 Test illegal behavior using multiple routers for the same transaction

jseyster committed:
SERVER-38623 Make safeNumberLongForHash consistent on all platforms

Previously, the hash function for double values relied on undefined
behavior when computing the hash for 2^63, meaning that it could give
different results on different platforms. Almost all platforms we
support returned -2^63 as the result of safeNumberLongForHash when the
input was 2^63, so this patch makes that the official result. Now,
safeNumberLongForHash will explicitly convert 2^63 to -2^63 on _all_
platforms without invoking any undefined behavior.

Note that this does not change our guidance that using a hashed index
for a field containing a floating point value greater than 2^53 is
unsupported:
https://docs.mongodb.com/manual/core/index-hashed/#considerations

ldennis committed:
SERVER-39672: default to local if no read concern specified for a multi-statment transaction

kaloianm committed:
SERVER-40297 Make all TransactionCoordinatorDriver methods free functions

The TransactionCoordinatorDriver is not really a "driver", but just a
set of functions to perform asynchronous work. There isn't any state to
keep, so there is no need for them to be in a class.

jseyster committed:
SERVER-40089 $group optimized with DISTINCT_SCAN cannot use $$ROOT

The getExecutorDistinct() function is responsible for both creating an
executor for the distinct command and creating an executor for a
$group that has been optimized with a DISTINCT_SCAN (see commit
da63195). These two scenarios have different requirements for their
projection, and getExecutorDistinct() distinguished the two by
assuming any caller with an empty ({}) projection wanted the distinct
command projection.

However, a $first accumulator with $$ROOT requires the entire
document, so the logic that builds an optimized $group executor
generates an empty projection for this case as well. When that
happens, getExecutorDistinct() mistakenly chooses the projection that
the distinct command wants, and when the pipeline evaluates $$ROOT, it
only gets to see a small subset of fields in the document.

This patch modifies getExecutorDistinct() so that the caller must
explicitly state what projection it wants. That means that the
distinct command no longer passes an empty projection to indicate that
it wants to project on just the distinct field. Instead, the distinct
command computes the projection for the distinct field on its own and
includes that projection in the ParsedDistinct object that it passes
to getExecutorDistinct().

jasonjhchan committed:
Revert "SERVER-40247 Add new replica_sets_multi_oplog_txns test suite for large transactions"

This reverts commit f8063475ecfde8a843cb701853c122a08433d9e8.

visemet committed:
SERVER-40241 Log resmoke.py invocation for local usage.

visemet committed:
Revert "SERVER-39899 Enable the initial sync fuzzer in Evergreen"

This reverts commit 6b8b02e8d72c055c6f686a781e53cde9384461b3.

visemet committed:
SERVER-39672 Fix lint.

smani87 committed:
SERVER-40008 Storage Interface changes to specify rounding off for prepare and commit timestamps.

smani87 committed:
SERVER-39782 Replaying prepared transaction writes for commit oplog entry during
    replication recovery and initial sync should set commit and durable
    timestamp for those writes.

Commit:
SERVER-39667: Locally disable flow control when majority reads are off.

banarun committed:
SERVER-40343 Better handling of errors from PCRE in $regex expressions

Commit:
SERVER-39870: Guarantee the oldest timestamp obeys any initial data timestamp constraints.

jasonjhchan committed:
SERVER-40247 Add new replica_sets_multi_oplog_txns test suite for large transactions

louiswilliams committed:
SERVER-40348 reindex_secondary.js should wait for index builds to complete on the secondary

louiswilliams committed:
SERVER-40349 buildindexes_false_with_system_indexes should wait for index builds to finish on secondaries

kaloianm committed:
SERVER-37881 Add a deadline for the coordinator preparing a transaction

Commit:
Revert "SERVER-39870: Guarantee the oldest timestamp obeys any initial data timestamp constraints."

This reverts commit ef38b62f0de257cfa0c5a2940c67949b7402079b.

GWlodarek committed:
SERVER-39593 Persist sizeStorer information before opening the backup cursor

sgolemon-corp committed:
SERVER-40373 Reduce race condition in test getlog2

acmorrow committed:
SERVER-40191 Use gperftools 2.7 on Ubuntu 18.04 ppc64le

acmorrow committed:
SERVER-36024 Relax overzealous dassert in mutable bson

BillyDonahue committed:
SERVER-40389 add fmt to third_party

lankas committed:
SERVER-37023 Test that initial sync applies operations from a transaction in separate storage transactions

vrachev committed:
SERVER-39899 Enable the initial sync fuzzer in Evergreen

ldennis committed:
SERVER-39310: Call checkCanServeReadsFor() in 'getMore'

- added call to checkCanServeReadsFor() in getmore_cmd.cpp after getting readlocks
- introduced two fail points: 1. pause 'getMore' before readlocks 2. pause rollback after state transition
- added testcase read_operations_during_rollback.js

vessy-mongodb committed:
SERVER-39989 Remove oldestActiveOplogEntryOpTime from serverStatus.transactions

Commit:
SERVER-39948 Remove simultaneous index builds specific curOp logging

jameswahlin committed:
SERVER-40310 Accept 'seconds' as an alias for 'secs' for the sleep command

jimoleary committed:
SERVER-38914 fix evergreen expansion issue

benety committed:
Revert "SERVER-40069 Fix global lock tracking for txns"

This reverts commit d99db775dbf73e27bd20e402f26f068dc199203d.

xy24 committed:
SERVER-39512 Make Database class thread-safe

bcaimano committed:
SERVER-40387 Shutdown shard registry after catalog cache

xy24 committed:
SERVER-39514 Move CollectionCatalogEntry ownership to UUIDCatalog

Commit:
SERVER-40213 Test upgrade/downgrade for $$NOW/$$CLUSTER_TIME in a view definition

Commit:
SERVER-40391 Remove low value explain2.js test

vessy-mongodb committed:
SERVER-40316 Wait for index build to complete in secondary_reads_timestamp_visibility.js

dstorch committed:
SERVER-40319 Don't call get() on boost::none in killCursorsWithMatchingSessions().

Schubes committed:
SERVER-37774 Platform Support: Remove Enterprise Ubuntu 16.04 PPCLE

jannaerin committed:
SERVER-39842 Allow update to shard key value if retryable write

spencerjackson committed:
SERVER-40394 Exempt buildindexes_false_with_system_indexes from fatal RoleGraph invalidation

lankas committed:
SERVER-40380 Test that inserting with a unique index and a prepared transaction causes prepare conflicts

ajdavis committed:
SERVER-36494 Test that active txn entries aren't truncated

Add tests for initial sync, recovery, and the inMemory storage engine.

Also, avoid taking a global X lock in replSetReconfig, we only need IX.

ajdavis committed:
SERVER-36494 Test that active txn entries aren't truncated

Add tests for initial sync, recovery, and the inMemory storage engine.

Also, avoid taking a global X lock in replSetReconfig, we only need IX.

visualzhou committed:
SERVER-40504 Check for interruption before checking for stepdown in awaitReplication()

spencerjackson committed:
SERVER-40385 Prevent collMod with usePowerOf2Sizes from invalidating the RoleGraph

visualzhou committed:
SERVER-40508 Remove hash from OplogSlot and make OplogSlot an alias for OpTime.

Commit:
SERVER-40005 Added setters for query and update to FindAndModifyRequest

This is necessary for a corresponding change in enterprise modules.
Note that recreating a request member by member just for FLE would
lead to defects when somebody adds new members at some point.

lukech committed:
Import wiredtiger: bb8086c4ed645854f19554198f391ea8c821f603 from branch mongodb-4.2

ref: 67066a329d..bb8086c4ed
for: 4.1.10

WT-4614       Trigger sweep immediately to reclaim space after a table drop
WT-4631       Always clear the read timestamp for transactions
WT-4636       Fix strace in syscall test
WT-4650       Coverity complaint: Constant expression result (CONSTANT_EXPRESSION_RESULT)
WT-4668       lint

Commit:
SERVER-40371 sys-perf: Add benchrun tests to new SSL variant

Commit:
SERVER-39974 Move setting index build curOp information from the MultiIndexBlock to the IndexBuildsCoordinator

ldennis committed:
SERVER-38257: Make transactionMetricsObserver.onAbort self-contained

- onAbort() calls _onAbortActive() if _singleTransactionStats.isActive(), otherwise it calls _onAbortInactive().
- add test case

Commit:
SERVER-40306 Ensure dropDatabase sets drop-pending back to false on failure

BillyDonahue committed:
SERVER-40389 small tweaks to third_party/fmt & related.

    Add an option to scons: --use-system-fmt
    import.sh: use the upstream libfmt github repo.
    StringData fmt interop
    noexcept for to_string_view(StringData) hook
    noexcept for StringData's rawData() and size()
    add fmt to THIRD-PARTY-NOTICES + markdown tweaks
    import the fmt LICENSE.rst
    update fmt lib revision

markbenvenuto committed:
SERVER-40330 clang_format.py format-my does not work with enterprise repo

pvselvan committed:
Revert "SERVER-35811 disallow committing at the prepareTimestamp and pin stable TS before oldest uncommitted TS"

This reverts commit b58420e57a2149c350ebf82815fe708850402296 and 33ac1afd4079e04d12554f9b79d1ab07426caf59.

benety committed:
SERVER-35811 fix lint

spencerjackson committed:
SERVER-40242 Add banner to MSI

kaloianm committed:
SERVER-40297 Make the TransactionCoordinator entirely futures-based

cswanson310 committed:
SERVER-40427 Update unittests_query.yml

Commit:
SERVER-38139 prepareTransaction will error if transaction operations were done against temporary collections

benety committed:
SERVER-39831 fix compile

benety committed:
SERVER-39831 remove failing test cases from OplogFetcherTest

Commit:
SERVER-39831 Add DEATH_TEST to OplogFetcherTest

ldennis committed:
SERVER-39690: Fix prepared transaction metrics after exceptions in prepare

- Also added unit test

Commit:
SERVER-39831 ephemeralForTest requires majority read concern

pvselvan committed:
SERVER-39042 Coordinator does not add one to commitTimestamp

louiswilliams committed:
SERVER-39260 Uncomment test at the bottom of  jstests/core/txns/timestamped_reads_wait_for_prepare_oplog_visibility.js

BlakeIsBlake committed:
SERVER-40350 Create a ReplOperation/Durable class delineation to indicate fields not persisted to disk

smani87 committed:
SERVER-40443 Transaction coordinator need to use w:majority instead of w:internalMajorityNoSnapshot.

visemet committed:
SERVER-40420 Clean up failed lock requests when using failpoint.

Using the "failNonIntentLocksIfWaitNeeded" failpoint with ResourceLock
would otherwise lead to invariant failure when destructing the Locker.

GWlodarek committed:
SERVER-39902 lockInfo should use UUIDCatalog to map resourceIds to collection names

TedTuckman committed:
SERVER-40243 Change EncryptionPlaceholder algorithm to be integer enum

smani87 committed:
SERVER-39041 Revert "SERVER-38094 Add write concern that waits for majority commit point but not committed snapshot"

This reverts commit 922e2f27c087cc017bb612a5c86eedb2cfa52a75.

henrikedin committed:
SERVER-40444 IndexCatalogEntry to use NamespaceString instead of plain string. This avoids converting between types later on.

markbenvenuto committed:
SERVER-40388 Fix Windows Push task for mongocryptd

BillyDonahue committed:
SERVER-40450 gperftools to use mongodb-forks org

kaloianm committed:
SERVER-39350 Make `opMsgRequestFromAnyProtocol` return owned request messages

bcaimano committed:
SERVER-39814 Bumping error codes to a reasonable band

20000 is probably going to be an error code eventually

GWlodarek committed:
SERVER-39231 Remove $nExtents from tests

GWlodarek committed:
SERVER-39231 'create' command should reject incorrect types for known options

benety committed:
Revert "SERVER-39814 Bumping error codes to a reasonable band"

This reverts commit 247a61781c6628e1fba1393f82ba8d7befb6334d.

benety committed:
Revert "SERVER-39814 Add OutOfLineExecutor to ConnectionPool factory interface"

This reverts commit f4b07839c904a73d28e3994af5bc902004fd4f9d.

benety committed:
Revert "SERVER-39965 OutOfLineExecutor Tasks are now unique_function(Status)"

This reverts commit 04ea1d46eb6c4c78e19409f120ae2e61f2a35204.

judahschvimer committed:
SERVER-39777 ensure nodes cannot step down during shutdown validation

jasonjhchan committed:
SERVER-40449 Add count field to commitTransaction for unprepared txns using large txns oplog format

Schubes committed:
SERVER-37773 Platform Support: Remove Community Ubuntu 16.04 ARM64

Schubes committed:
SERVER-37765 Platform Support: Remove Ubuntu 14.04

markbenvenuto committed:
SERVER-40393 Disable SSL_MODE_RELEASE_BUFFERS in ASIO

lankas committed:
SERVER-40018 Remove ServerTransactionMetrics::getOldestActiveOpTime and supporting datastructures

vrachev committed:
SERVER-40421 Add failpoint to skip doing retries on WiredTiger prepare conflicts

BillyDonahue committed:
SERVER-40476 remove mongoutils::str::equals

Commit:
percentile unittests added

GWlodarek committed:
SERVER-40345 Change calls involving the 'temp' field in the 'create' command to be done through the 'applyOps' command

GWlodarek committed:
SERVER-40345 The 'temp' field in the 'create' command is only allowed to be used internally and isn't available to clients

judahschvimer committed:
SERVER-38937 unify txn override and auto retry on network error

briananthonymccarthy committed:
SERVER-40491 Update Debian 8 image for package test

louiswilliams committed:
SERVER-40291 Wait for index builds to complete on secondaries at the end of apply_ops_create_indexes

Commit:
SERVER-40312 Create a generic tree for pipeline metatdata

GWlodarek committed:
SERVER-40481 Fix data race in UUIDCatalog::iterator::iterator()

GWlodarek committed:
SERVER-39479 Increase number of executions in 'Linux Repeated Executions' variant

ldennis committed:
SERVER-40337: Remove oldestOpenUnpreparedReadTimestamp from serverStatus.transactions

This reverts most of the work done in 3bb72042c29ceac96ec628ee643cbc4db8e982ab

henrikedin committed:
SERVER-32709 Fix duplicate error code

Commit:
SERVER-40312 Create a generic tree for pipeline metatdata

GWlodarek committed:
SERVER-40162 Change the default WiredTiger session_max from 20k to 33k

Commit:
SERVER-40312 Create a generic tree for pipeline metatdata

mattlord committed:
SERVER-39749 Name discrepancy for Linux arm64 download artifacts

chasinglogic committed:
SERVER-32295 Support Python 3 builds in Perf configs

jameswahlin committed:
SERVER-40497 Fix ViewCatalog call to get() on empty boost::optional

louiswilliams committed:
SERVER-40451 Remove diskFull.js

GWlodarek committed:
SERVER-40148 Load implicitly_retry_on_background_op_in_progress override in the checkReplicatedDataHashes thread

henrikedin committed:
SERVER-32709 Correctly pass cache_size to SQLite.

Commit:
Revert "SERVER-40312 Create a generic tree for pipeline metatdata"

This reverts commit aad70256024c8447344a4d7f1cfdda742d1904ee.
This reverts commit 3765a3200adb31239fad3abaeddebe81e5aa9280.
This reverts commit f16366d55edae0bf02cd3d0a532e31c7c938e8db.
This reverts commit 33598e48f9be86c16990051934d4e72f073c2b86.

louiswilliams committed:
SERVER-40175 Rebuild any missing _id indexes on startup

vessy-mongodb committed:
SERVER-40502 Turn off failpoint before shutting down test in ddl_ops_after_prepare_lock_failpoint.js

markbenvenuto committed:
SERVER-40112 db.disableFreeMonitoring() returns 'not master' on secondary after restart

benety committed:
SERVER-40169 OplogEntry constructor accepts prepare argument

benety committed:
SERVER-39676 add missing library dependency

chasinglogic committed:
SERVER-40514 Fix race in Resmoke

visualzhou committed:
SERVER-40336 Use SecureRandom to seed the random number generator in replication coordinator.

benety committed:
SERVER-40169 add unit test for OplogApplier with test cases for CRUD and applyOps operations

GWlodarek committed:
SERVER-39663 change dropIndex("*") to dropIndexes("*") to drop multiple indexes

GWlodarek committed:
SERVER-39663 db.collection.dropIndex() should not accept multiple index names

benety committed:
SERVER-40169 OplogApplier::getNextApplierBatch() processes system.views oplog entries individually

This applies changes from commit f78e580dad60fa6ece39f34b3dcfc3ed95b85414.

benety committed:
SERVER-40169 OplogApplier::getNextApplierBatch() processes admin.system.version oplog entries individually

This applies changes from commit 0d79175a88ee958722c0ffb276606949e695d028.

benety committed:
SERVER-39514 fix UUIDCatalog::next() to avoid advancing past end of map

benety committed:
SERVER-40169 OplogApplier::getNextApplierBatch() groups unprepared commitTransaction oplog entries with CRUD ops

This applies changes from commit 3328515f7d80c8cedcaf8c0df83c6effc60330d0.

benety committed:
SERVER-39950 OplogApplier::getNextApplierBatch() handles BatchLimit::slaveDelayLatestTimestamp

benety committed:
SERVER-39950 SyncTail::oplogApplication() accepts OplogApplier::GetNextApplierBatchFn

benety committed:
SERVER-39950 make OplogApplier::getNextApplierBatch() shutdown-aware

benety committed:
SERVER-39950 SyncTail::OpQueueBatcher::run() obtains next batch of ops using OplogApplier

benety committed:
SERVER-39950 remove SyncTail::tryPopAndWaitForMore() and _consume()

lukech committed:
Import wiredtiger: 1768d66613fc32b664ac3608a4e740d5e8c6fd0f from branch mongodb-4.2

ref: bb8086c4ed..1768d66613
for: 4.1.10

WT-4639       Use test/format to verify database file portability
WT-4644       column-store missing run-length encoding opportunities
WT-4652       Return an error if durable timestamp is not passed to a prepared transaction.
WT-4687       test/format can try to set oldest timestamp to 0
WT-4693       WT_CONNECTION::reconfigure should not require quiescence when downgraded

Commit:
SERVER-39717: Protect against overflow in flow control calculations.

xy24 committed:
SERVER-39515 Remove DatabaseCatalogEntry

Commit:
SERVER-40415 Tempfile cleanup from test_adb_monitor.py

saltzm committed:
SERVER-40396 Remove invalid check from moveChunk concurrency workload

BillyDonahue committed:
SERVER-40389 fixup: fmtlib.net url has no www in it

jasonjhchan committed:
SERVER-39793 Update the state in transaction table for running transactions on secondary

will62794 committed:
SERVER-39660 Move readSource re-definition to a clearer location in AutoGetCollectionForRead

ajdavis committed:
Revert "SERVER-36494 Test that active txn entries aren't truncated"

This reverts commit f8f872e029ba3b1f32d8499c912756d48dc1a03b.

jannaerin committed:
SERVER-40318 Make condition variable wait in NamespaceSerializer exception safe

chasinglogic committed:
SERVER-32295 Remove references to Python 2

benety committed:
SERVER-39436 fix batch limit checking in OplogApplier::getNextApplierBatch()

benety committed:
SERVER-39436 check batch limits on large unprepared transactions using embedded operation count

BillyDonahue committed:
SERVER-40476 remove mongoutils::str

    Rename utils/mongoutils/str.h => utils/str.h
    Rename namespace mongoutils::str => str
    Rename mongo::strcasecmp => str::caseInsensitiveCompare.

mtrussotto committed:
SERVER-39790 Reconstruct prepared transactions from new oplog format on startup

will62794 committed:
SERVER-39969 Update the 'mongod-dump-sessions' GDB command to be compatible with the new internal format of TransactionParticipant

visemet committed:
SERVER-40532 Install Python dependencies in more places.

These callers of Python failing to install their own dependencies had
gone unnoticed because the "check binary version" function would have
incidentally done it for them. However, the "check binary version"
function is a no-op when bypassing compile due to no C++ changes being
made.

visemet committed:
SERVER-40518 Change backup_restore*.js tests to send resmoke.py SIGINT.

ldennis committed:
SERVER-40269: commitTransaction should assert that the prepare oplog entry is majority committed

jasonjhchan committed:
SERVER-40331 Add commit and abort jstests for transactions larger than 16MB

acmorrow committed:
SERVER-40375 Remove gperftools 2.5

jasonjhchan committed:
SERVER-39829 Consider in-progress transactions when calculating the oldest active transaction timestamp

Commit:
SERVER-39848: Add flow control diagnostics.

markbenvenuto committed:
SERVER-40542 buildscripts\make_vcxproj.py does not work with Python 3

markbenvenuto committed:
SERVER-40543 gen_casefold_map.py does not work with CRLF line endings

ldennis committed:
SERVER-40329: Log all transactions at TXN verbosity levels >= 1

Commit:
fix definition bug in header

stephan-hof commented:
Looks like github has issues with 'rebase' and 'force-push'.
I'll close this one and and create a new one.
